Bug: Executing AndroidTests fails because connectedDebugAndroidTest task fails to build

Executing any test inside of the androidTest directory leads to a build-exception. 

```
> Task :compileDebugJavaWithJavac
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

> Task :connectedDebugAndroidTest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':connectedDebugAndroidTest'.
```

The current analysis shows that this Exception may be caused by an internal module that throws an `IllegalAccessError`.

<img width="1630" alt="image" src="https://github.com/user-attachments/assets/aef29831-e988-4be1-a806-d1e6af23328d">

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -14,5 +14,11 @@ jobs:
       - name: Checkout
         uses: "actions/checkout@v3"
 
+      - name: Set up Java
+        uses: actions/setup-java@v2
+        with:
+          java-version: 17
+          distribution: "temurin"
+
       - name: Build with Gradle
         run: ./gradlew build

--- a/build.gradle
+++ b/build.gradle
@@ -8,7 +8,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.4.2'
+        classpath "com.android.tools.build:gradle:8.6.1"
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
     }
 }
@@ -35,6 +35,10 @@ android {
     compileSdkVersion 34
     namespace "org.billthefarmer.editor"
 
+    buildFeatures {
+        buildConfig = true
+    }
+
     defaultConfig {
         applicationId "org.billthefarmer.editor"
         minSdkVersion 21
@@ -52,8 +56,8 @@ android {
 
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_11
-        targetCompatibility JavaVersion.VERSION_11
+        sourceCompatibility JavaVersion.VERSION_17
+        targetCompatibility JavaVersion.VERSION_17
     }
 
     lintOptions {
@@ -71,7 +75,7 @@ android {
 
     packagingOptions { resources.excludes.add("META-INF/*") }
     kotlinOptions {
-        jvmTarget = '11'
+        jvmTarget = '17'
     }
     buildFeatures {
         compose true
@@ -81,7 +85,7 @@ android {
 task jacocoTestReport(type: JacocoReport, dependsOn: 'testDebugUnitTest') {
 
     reports {
-        xml.enabled true
+//        xml.enabled true
     }
 
     def fileFilter = ['**/R.class', '**/R$*.class', '**/BuildConfig.*', '**/Manifest*.*', '**/*Test*.*', 'android/**/*.*']

--- a/build.gradle
+++ b/build.gradle
@@ -107,49 +107,45 @@ sonarqube {
 }
 
 dependencies {
+    // Used by the Editor
     implementation 'org.commonmark:commonmark:0.22.0'
 
+    // Android Compose for Kotlin Compiler
+    implementation 'androidx.compose.runtime:runtime:1.7.5'
 
-    implementation 'org.junit.jupiter:junit-jupiter:5.11.1'
-    implementation 'androidx.core:core-ktx:1.3.0'
-    implementation 'androidx.test.ext:junit:1.1.1'
-//    implementation 'androidx.lifecycle:lifecycle-runtime-ktx:2.6.1'
-//    implementation 'androidx.activity:activity-compose:1.4.0'
-//    implementation platform('androidx.compose:compose-bom:2022.10.00')
-//    implementation 'androidx.compose.runtime:runtime-dispatch:1.0.0-alpha12'
-    implementation 'androidx.compose.runtime:runtime:1.0.1'
+    implementation 'androidx.core:core-ktx:1.12.0'
+    implementation 'androidx.appcompat:appcompat:1.6.1'
+    implementation 'com.google.android.material:material:1.10.0'
+    implementation 'androidx.constraintlayout:constraintlayout:2.1.4'
+    implementation 'androidx.navigation:navigation-ui-ktx:2.7.4'
 
+    // Local Tests
+    testImplementation 'junit:junit:4.13.2'
+    testImplementation 'androidx.arch.core:core-testing:2.2.0'
+    testImplementation 'io.mockk:mockk:1.13.8'
+    testImplementation "com.google.truth:truth:1.2.0"
+    testImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.8.0-RC2'
 
+    // Instrumented Tests
+    // Kotlin Coroutines testing support (we added this in the last tutorial)
+    androidTestImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.8.0-RC2'
 
-//    implementation 'androidx.compose.ui:ui'
-//    implementation 'androidx.compose.ui:ui-graphics'
-//    implementation 'androidx.compose.ui:ui-tooling-preview'
-//    implementation 'androidx.compose.material3:material3'
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.11.1'
-    testImplementation 'org.mockito:mockito-core:5.14.0'
-    testImplementation("org.junit.jupiter:junit-jupiter-params:5.11.1")
-    testImplementation 'junit:junit:4.12'
-//    androidTestImplementation platform('androidx.compose:compose-bom:2022.10.00')
-//    androidTestImplementation 'androidx.compose.ui:ui-test-junit4'
-//    debugImplementation 'androidx.compose.ui:ui-tooling'
-//    debugImplementation 'androidx.compose.ui:ui-test-manifest'
+    // Since this needs to run on an Android device we must add this one to get the RunWith(AndroidJUnit4::class)
+    androidTestImplementation "androidx.test.ext:junit:1.1.5"
 
-    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.11.1'
+    // To test LiveData we also need Arch Core
+    androidTestImplementation "androidx.arch.core:core-testing:2.2.0"
 
-    testImplementation 'androidx.test.espresso:espresso-core:3.4.0'
-    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
+    // Assertion library for more readable assertions
+    androidTestImplementation "com.google.truth:truth:1.2.0"
 
-    testImplementation 'androidx.test:rules:1.5.0'
-    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    // Espresso Core - needed for AndroidJUnit4 runner
+    androidTestImplementation "androidx.test.espresso:espresso-core:3.5.1"
 
-
-    // Espresso dependencies
-    androidTestImplementation 'androidx.test.espresso:espresso-intents:3.4.0'
-
-    // AndroidX Test dependencies
-    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
-    androidTestImplementation 'androidx.test:runner:1.4.0'
-    androidTestImplementation 'androidx.test:rules:1.4.0'
+    // Lifecycle
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.6.2"
+    implementation "androidx.lifecycle:lifecycle-livedata-ktx:2.6.2"
+    implementation "androidx.lifecycle:lifecycle-extensions:2.2.0"
 
 }
 

--- a/build.gradle
+++ b/build.gradle
@@ -32,13 +32,13 @@ apply plugin: 'org.sonarqube'
 apply plugin: 'com.android.application'
 
 android {
-    compileSdkVersion 30
+    compileSdkVersion 34
     namespace "org.billthefarmer.editor"
 
     defaultConfig {
         applicationId "org.billthefarmer.editor"
         minSdkVersion 21
-        targetSdkVersion 28
+        targetSdkVersion 30
         versionName "1.96"
         versionCode 196
 
@@ -139,9 +139,18 @@ dependencies {
     testImplementation 'androidx.test.espresso:espresso-core:3.4.0'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
 
-    androidTestImplementation 'androidx.test:runner:1.5.0'
     testImplementation 'androidx.test:rules:1.5.0'
     implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+
+
+    // Espresso dependencies
+    androidTestImplementation 'androidx.test.espresso:espresso-intents:3.4.0'
+
+    // AndroidX Test dependencies
+    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
+    androidTestImplementation 'androidx.test:runner:1.4.0'
+    androidTestImplementation 'androidx.test:rules:1.4.0'
+
 }
 
 

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,2 +1,3 @@
 org.gradle.jvmargs=-Xmx1536M
 android.useAndroidX=true
+android.nonFinalResIds=false

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,7 +1,7 @@
-#Fri Nov 01 11:26:49 CET 2024
+#Sat Dec 07 11:20:53 CET 2024
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-bin.zip
 networkTimeout=10000
 validateDistributionUrl=true
 zipStoreBase=GRADLE_USER_HOME

--- a/src/androidTest/kotlin/editor/KotlinIntegrationTest.kt
+++ b/src/androidTest/kotlin/editor/KotlinIntegrationTest.kt
@@ -7,14 +7,14 @@ import androidx.test.ext.junit.runners.AndroidJUnit4
 import org.billthefarmer.editor.Editor
 import org.billthefarmer.editor.R
 import org.junit.Rule
-import org.junit.jupiter.api.Test
+import org.junit.Test
 import org.junit.runner.RunWith
 
 @RunWith(AndroidJUnit4::class)
 class KotlinIntegrationTest {
 
-//    @get:Rule
-//    val activityRule = ActivityScenarioRule(Editor::class.java)
+    @get:Rule
+    val activityRule = ActivityScenarioRule(Editor::class.java)
 
 
     @Test

--- a/src/androidTest/kotlin/editor/KotlinIntegrationTest.kt
+++ b/src/androidTest/kotlin/editor/KotlinIntegrationTest.kt
@@ -1,20 +1,26 @@
 
+import android.view.View
 import androidx.test.espresso.Espresso.onView
 import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.action.ViewActions.replaceText
 import androidx.test.espresso.assertion.ViewAssertions.matches
 import androidx.test.espresso.matcher.ViewMatchers.isClickable
-import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
-import androidx.test.espresso.matcher.ViewMatchers.isNotClickable
+import androidx.test.espresso.matcher.ViewMatchers.isFocusable
 import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.espresso.matcher.ViewMatchers.withResourceName
 import androidx.test.espresso.matcher.ViewMatchers.withText
 import androidx.test.ext.junit.rules.ActivityScenarioRule
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import org.billthefarmer.editor.Editor
 import org.billthefarmer.editor.R
-import org.hamcrest.core.IsNot.not
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers.allOf
+import org.hamcrest.TypeSafeMatcher
 import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
+import java.lang.Thread.sleep
 
 @RunWith(AndroidJUnit4::class)
 class KotlinIntegrationTest {
@@ -24,20 +30,29 @@ class KotlinIntegrationTest {
 
     @Test
     fun testCreateAndSaveFile_WithContent(){
+        val TEXT_INPUT = "Hello from this test-case!"
+        val FILE_PATH = "Documents/Example-File.txt"
 
-        // Verify pre-conditions edit visible, view invisible
-        onView(withId(R.id.view)).check(matches((isNotClickable())))
-        onView(withId(R.id.edit)).check(matches((isClickable())))
+        onView(withId(R.id.newFile)).check(matches((isClickable())))
+        onView(withId(R.id.newFile)).perform(click())
 
+        onView(withId(R.id.text)).perform(click(), replaceText(TEXT_INPUT))
+        onView(withId(R.id.text)).check(matches(withText(TEXT_INPUT)))
 
-        // Click edit icon to disable read-only mode
-        onView(withId(R.id.edit)).perform(click())
+        onView(withId(R.id.save)).check(matches(isClickable()))
+        onView(withId(R.id.save)).perform(click())
 
-        onView((withId(R.id.edit))).check(matches(withText("")))
-
-        // Edit button should now be invisible
-        onView(withId(R.id.edit)).check(matches(isNotClickable()))
-        onView(withId(R.id.view)).check(matches(isClickable()))
+        // Define Save Location
+        onView(withId(R.id.pathText)).perform(replaceText(FILE_PATH))
+        onView(
+           allOf(
+               withText("Save"),
+               withResourceName("button1") // identifier of the save-button of file-save-dialog.
+           )
+        )
+            .check(matches(isClickable()))
+            .check(matches(isFocusable()))
+            .perform(click())
     }
 
 

--- a/src/androidTest/kotlin/editor/KotlinIntegrationTest.kt
+++ b/src/androidTest/kotlin/editor/KotlinIntegrationTest.kt
@@ -1,11 +1,17 @@
 
 import androidx.test.espresso.Espresso.onView
 import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.isClickable
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.isNotClickable
 import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.espresso.matcher.ViewMatchers.withText
 import androidx.test.ext.junit.rules.ActivityScenarioRule
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import org.billthefarmer.editor.Editor
 import org.billthefarmer.editor.R
+import org.hamcrest.core.IsNot.not
 import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
@@ -15,6 +21,24 @@ class KotlinIntegrationTest {
 
     @get:Rule
     val activityRule = ActivityScenarioRule(Editor::class.java)
+
+    @Test
+    fun testCreateAndSaveFile_WithContent(){
+
+        // Verify pre-conditions edit visible, view invisible
+        onView(withId(R.id.view)).check(matches((isNotClickable())))
+        onView(withId(R.id.edit)).check(matches((isClickable())))
+
+
+        // Click edit icon to disable read-only mode
+        onView(withId(R.id.edit)).perform(click())
+
+        onView((withId(R.id.edit))).check(matches(withText("")))
+
+        // Edit button should now be invisible
+        onView(withId(R.id.edit)).check(matches(isNotClickable()))
+        onView(withId(R.id.view)).check(matches(isClickable()))
+    }
 
 
     @Test

--- a/src/main/AndroidManifest.xml
+++ b/src/main/AndroidManifest.xml
@@ -5,9 +5,10 @@
     android:installLocation="auto"
     tools:ignore="GoogleAppIndexingWarning">
 
-  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
+  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"  android:maxSdkVersion="30"/>
 
   <application
+      android:requestLegacyExternalStorage="true"
       android:allowBackup="true"
       android:icon="@drawable/ic_launcher"
       android:label="@string/appName"

--- a/src/test/java/example/ExampleTest.java
+++ b/src/test/java/example/ExampleTest.java
@@ -2,12 +2,13 @@ package example;
 
 import static org.junit.Assert.*;
 
-import org.junit.jupiter.api.Test;
+import org.junit.Test;
 
-class ExampleTest {
+
+public class ExampleTest {
 
     @Test
-    void checkNickname() {
+    public void checkNickname() {
         assertTrue(true);
     }
 

